src/senselab/audio/tasks/features_extraction/praat_parselmouth.py

```diff
@@ -1,143 +1,1377 @@
+                        "Robust",
+                    )
+                    # Subtract tilt before smoothing = “no”; time averaging window = 0.01 s;
+                    # quefrency averaging window = 0.001 s;
+                    # Peak search pitch range = 60–330 Hz; tolerance = 0.05; interpolation = “Parabolic”;
+                    # tilt line frequency range = 0.001–0 s (no upper bound);
+                    # Line type = “Straight”; fit method = “Robust.”
+                except Exception as e:
+                    current_frame = inspect.currentframe()
+                    if current_frame is not None:
+                        current_function_name = current_frame.f_code.co_name
+                        logger.error(f'Error in "{current_function_name}": \n' + str(e))
+                    CPP_Value = np.nan
+
+                if not np.isnan(CPP_Value) and CPP_Value > 4:
+                    cpp_list.append(CPP_Value)
+
+        # Calculate Final Features
+        if cpp_list:
+            CPP_array = np.array(cpp_list)
+            CPP_mean = np.mean(CPP_array)
+            CPP_std = np.std(CPP_array)
+        else:
+            CPP_mean = np.nan
+            CPP_std = np.nan
+
+        # Return Result
+        return {"mean_cpp": CPP_mean, "std_dev_cpp": CPP_std}
+
+    except Exception as e:
+        current_frame = inspect.currentframe()
+        if current_frame is not None:
+            current_function_name = current_frame.f_code.co_name
+            logger.error(f'Error in "{current_function_name}": \n' + str(e))
+        return {"mean_cpp": float("nan"), "std_dev_cpp": float("nan")}
+
+
+def measure_formants(
+    snd: Union[parselmouth.Sound, Path, Audio], floor: float, ceiling: float, frame_shift: float
+) -> Dict[str, float]:
+    """Extract Formant Frequency Features.
+
+    Function to extract formant frequency features from a given sound object. This function is adapted from default
+    Praat code to work with Parselmouth.
+
+    Args:
+        snd (Union[parselmouth.Sound, Path, Audio]): A Parselmouth Sound object or a file path or an Audio object.
+        floor (float): Minimum expected pitch value, set using value found in `pitch_values` function.
+        ceiling (float): Maximum expected pitch value, set using value found in `pitch_values` function.
+        frame_shift (float): Time rate at which to extract a new pitch value, typically set to 5 ms.
+
+    Returns:
+        dict: A dictionary containing the following keys:
+
+            - f1_mean (float): Mean F1 location.
+            - f1_std (float): Standard deviation of F1 location.
+            - b1_mean (float): Mean F1 bandwidth.
+            - b1_std (float): Standard deviation of F1 bandwidth.
+            - f2_mean (float): Mean F2 location.
+            - f2_std (float): Standard deviation of F2 location.
+            - b2_mean (float): Mean F2 bandwidth.
+            - b2_std (float): Standard deviation of F2 bandwidth.
```

Comment:
is it only f1 and f2? if so, perhaps rename function to `measure_f1f2_formants_bandwidths`